### DISCUSSION TITLE
Bundle custom i18n translations with extensions

### DISCUSSION THREAD

**Discussion by programmarchy** [directus/directus#19229] [Feature Requests]
### Summary

Although extensions can use _existing_ i18n translations, there doesn't seem to be a good way to define new custom translations that get loaded with the extension. This would be especially useful for app and hybrid extensions with interface fields so that translations could be used for field labels, field notes, and other custom UI text. For example, built-in operations use translations to [define an operation's name and description](https://github.com/directus/directus/blob/main/app/src/operations/sleep/index.ts#L6) and it'd be great if extensions could do this, too.

### Basic Example

When an extension is loaded, load `translations/*.yaml` from the extension package the same way Directus loads its built in translations.

### Motivation

Directus is a Multilingual App which supports full internationalization.

However, extensions cannot currently leverage Directus's i18n system to its full capacity.

### Detailed Design

Update the `create-directus-extension` tool to copy `src/lang/translations/*.yaml` to the `dist` folder upon build.

Modify [`setLanguage`](https://github.com/directus/directus/blob/main/app/src/lang/set-language.ts#L9) to iterate through loaded extensions and also merge in any translations by extensions.

### Must Have:
- Translations must be loaded before any UI components defined by the extension are displayed.

### Should Have:
- Use the same architecture currently used to load built in Directus translations.

### Could Have:
- Translations for extensions could be namespaced to prevent conflicts with other extensions or Directus itself.

### Won't Have:
- N/A

### Drawbacks

- Adds complexity to the i18n system. Might be a difficult change if `setLanguage` is called before extensions are loaded in the Directus app lifecycle.
- Should `EXTENSIONS_AUTO_RELOAD` detect changes to translations, too?
- Could allow extension authors to override built-in translations.
- Could cause conflicts between multiple extensions if translations aren't namespaced.

### Alternatives

Perhaps one workaround might be to define an `init` hook extension that uses the translation service to install any necessary translations, and include it in a bundle?

### Adoption Strategy

Adoption would be opt-in. Extension developers would simply include the relevant translations in their extension.

### Unresolved Questions

Would translations defined in extensions be able to override built in Directus translations? User-defined translations? In what order would translations be loaded if there are multiple extensions? Should order matter?

**Comment by nawafinity** [comment]
This feature would be incredibly valuable for making **Directus extensions** support global languages natively.

In the meantime, I’ve found a workaround that allows you to expose and use your own custom translations within your extension. This approach may be useful for anyone building a custom module or interface that needs multi-language support.

### 🌍 How to Add Multi-Language Support in a **Directus Extension**

1. **Install `vue-i18n`**
2. **Create a `resolveLocale` Utility**

This function normalizes locale strings to match your translation keys (e.g., converting `en-US` to `en`).

```ts
export function resolveLocale(locale: string): string {
  if (locale) {
    if (locale.includes('-')) {
      return locale.split('-')[0].toLowerCase();
    }
    return locale.toLowerCase();
  }
  return 'en';
}
```

3. **Access the Default Language from Directus**

You can use the `settingsStore` to get the currently active language in the Directus Data Studio.

```ts
import { useStores } from '@directus/extensions-sdk';

const { useSettingsStore } = useStores();
const { settings } = useSettingsStore();
const defaultLanguage = settings?.default_language;
```

4. **Initialize `vue-i18n` Locally Using the Composition API**

You can scope translations within your component like this:

```ts
import { useI18n } from 'vue-i18n';
import { resolveLocale } from '../shared/utils/i18n';

const { t } = useI18n({
  locale: resolveLocale(defaultLanguage),
  messages: {
    en: {
      zme_please_wait: 'Please wait...',
    },
    ar: {
      zme_please_wait: 'يرجى الانتظار...',
    },
  },
});
```

> 💡 By doing this, you can define and use your own localized strings, or even use [[Directus core translation keys](https://github.com/directus/directus/tree/main/app/src/lang)](https://github.com/directus/directus/tree/main/app/src/lang) if needed.

I hope this workaround proves helpful to others looking to add multi-language support to their Directus extensions. Until native i18n support for extensions is introduced, this approach offers a flexible and reliable solution.
